fiat-ldap/src/main/java/com/netflix/spinnaker/fiat/roles/ldap/LdapUserRolesProvider.java

```diff
@@ -93,12 +99,55 @@
         .collect(Collectors.toList());
   }
 
+  private class GroupToUsers implements AttributesMapper<Pair<String, List<String>>> {
+    public Pair<String, List<String>> mapFromAttributes(Attributes attrs) throws NamingException {
+      String group = attrs.get(configProps.getGroupRoleAttributes()).get().toString();
+      NamingEnumeration<?> members = attrs.get(configProps.getGroupsUserAttributes()).getAll();
+      List<String> member = new ArrayList<>();
+      while (members.hasMore()) {
+        try {
+          String user =
+              Arrays.stream(configProps.getUserDnPattern().parse(members.next().toString()))
+                  .toArray(String[]::new)[0];
+          member.add(user);
+        } catch (ParseException e) {
+          e.printStackTrace();
+        }
+      }
+      return Pair.of(group, member);
+    }
+  }
+
   @Override
   public Map<String, Collection<Role>> multiLoadRoles(Collection<ExternalUser> users) {
     if (StringUtils.isEmpty(configProps.getGroupSearchBase())) {
       return new HashMap<>();
     }
 
+    if (users.size() > configProps.getThresholdToUseGroupFilter()
+        && StringUtils.isNotEmpty(configProps.getGroupsFilter())) {
+      Map<String, Collection<Role>> userRolesMap =
+          users.stream()
+              .collect(Collectors.toMap(ExternalUser::getId, ExternalUser::getExternalRoles));
+      Set<String> userIds = userRolesMap.keySet();
+      ldapTemplate
+          .search(
+              configProps.getGroupSearchBase(), configProps.getGroupsFilter(), new GroupToUsers())
+          .forEach(
```

Comment:
I'm not with IDE in my hand right now but can we reduce the indentation level by doing something like

```
.flatMap(entry -> entry.getValue())
.filter(userIds::contains)
.forEach(...)
```

Reply:
Modified.